src/mono/mono/metadata/appdomain.c

```diff
@@ -1256,27 +1276,135 @@ mono_runtime_install_appctx_properties (void)
 	g_assert (setup);
 
 	// FIXME: TRUSTED_PLATFORM_ASSEMBLIES is very large
 
+	// Combine and convert properties
+	if (buffer)
+		n_runtimeconfig_json_props = mono_metadata_decode_value(buffer, &buffer);
+
+	n_total_props = n_appctx_props + n_runtimeconfig_json_props;
+	total_keys = g_new0 (gunichar2*, n_total_props);
+	total_values = g_new0 (gunichar2*, n_total_props);
+
+	for (int i = 0; i < n_appctx_props; ++i) {
+		total_keys [i] = g_utf8_to_utf16 (appctx_keys [i], strlen (appctx_keys [i]), NULL, NULL, NULL);
+		total_values [i] = g_utf8_to_utf16 (appctx_values [i], strlen (appctx_values [i]), NULL, NULL, NULL);
+	}
+
+	runtimeconfig_json_read_props (buffer, &buffer, n_runtimeconfig_json_props, total_keys+n_appctx_props, total_values+n_appctx_props);
+
 	/* internal static unsafe void Setup(char** pNames, char** pValues, int count) */
-	args [0] = appctx_keys;
-	args [1] = appctx_values;
-	args [2] = &n_appctx_props;
+	args [0] = total_keys;
+	args [1] = total_values;
+	args [2] = &n_total_props;
 
 	mono_runtime_invoke_checked (setup, NULL, args, error);
 	mono_error_assert_ok (error);
 
+	if (runtimeconfig_json_map != NULL) {
+		mono_file_unmap ((gpointer)buffer_start, runtimeconfig_json_map_handle);
+		mono_file_map_close (runtimeconfig_json_map);
+	}
+
+	// Call user defined cleanup function
+	if (runtime_config_cleanup_fn)
+		(*runtime_config_cleanup_fn) (runtime_config_arg, runtime_config_user_data);
+
 	/* No longer needed */
+	for (int i = 0; i < n_total_props; ++i) {
+		g_free (total_keys [i]);
+		g_free (total_values [i]);
+	}
+	g_free (total_keys);
+	g_free (total_values);
 	for (int i = 0; i < n_appctx_props; ++i) {
 		g_free (appctx_keys [i]);
 		g_free (appctx_values [i]);
 	}
 	g_free (appctx_keys);
 	g_free (appctx_values);
+
 	appctx_keys = NULL;
 	appctx_values = NULL;
+	if (runtime_config_arg){
+		runtime_config_arg = NULL;
+		runtime_config_cleanup_fn = NULL;
+		runtime_config_user_data = NULL;
+	}
+}
+
+static const char*
+runtimeconfig_json_get_buffer (MonovmRuntimeConfigArguments *arg, MonoFileMap **file_map, gpointer *buf_handle)
+{
+	if (arg != NULL)
+	{
+		switch (arg->kind) {
+		case 0: {
+			char *buffer = NULL;
+			guint64 file_len = 0;
+
+			*file_map = mono_file_map_open (arg->runtimeconfig.name.path);
+			g_assert (*file_map);
+			file_len = mono_file_map_size (*file_map);
+			g_assert (file_len > 0);
+			buffer = (char*)mono_file_map (file_len, MONO_MMAP_READ|MONO_MMAP_PRIVATE, mono_file_map_fd (*file_map), 0, buf_handle);
+			g_assert (buffer);
+			return buffer;
+		}
+		case 1: {
+			*file_map = NULL;
+			*buf_handle = NULL;
+			return arg->runtimeconfig.data.data;
+		}
+		default:
+			g_assert_not_reached ();
+		}
+	}
+
+	*file_map = NULL;
+	*buf_handle = NULL;
+	return NULL;	
+}
+
+static void
+runtimeconfig_json_read_props (const char *ptr, const char **endp, int nprops, gunichar2 **dest_keys, gunichar2 **dest_values)
```

Comment:
Looking a little more closely I think you can skip the utf8 intermediate buffers entirely and pass the decoded length to g_utf8_to_utf16 directly?

Reply:
Good point! Removed utf8 intermediate buffers entirely. Thanks!